### PR TITLE
Configuring Segment Heap for bpf2c binary 

### PR DESCRIPTION
**Pull request by shpalani**
## Description
Windows OS Heap dev and MSVC ASAN dev team suggested moving from NT Heap to Segment Heap to unblock the heap failures encountered in CI/CD pipeline build for FuzzerDebug in bpf2c binary. Details in #2404.

2 suggestions:
1. Have a manifest file. The manifest approach is per application binary.
2. Set registry. The registry approach is per machine setting.

This PR contains the first approach of embedding a manifest file containing 'SegmentHeap' in the bpf2c binary, in Debug/Release/FuzzerDebug builds.
https://learn.microsoft.com/en-us/windows/win32/sbscs/application-manifests#heaptype

## Testing

_Do any existing tests cover this change? No
Are new tests needed? No

Manual testing:
```
1. Loaded FuzzerDebug (C:\ebpf-2023\ebpf-for-windows\x64\FuzzerDebug\bpf2c.exe) and checked the heap
0:003> !heap
        Heap Address      NT/Segment Heap

         19cb47a0000         Segment Heap
         19cb46e0000              NT Heap
```
```
Loaded Debug image (C:\ebpf-2023\ebpf-for-windows\x64\Debug\bpf2c.exe) and checked the heap
0:000> !heap
        Heap Address      NT/Segment Heap

         19167430000         Segment Heap
         19167370000              NT Heap
```

```
C:\ebpf-2023\ebpf-for-windows>cd x64\Debug

C:\ebpf-2023\ebpf-for-windows\x64\Debug>type bpf2c.exe.manifest
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<!--
  Copyright (c) Microsoft Corporation
  SPDX-License-Identifier: MIT

  Overrides the default heap implementation with segment heap.
-->
<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0" xmlns:asmv3="urn:schemas-microsoft-com:asm.v3">
  <assemblyIdentity name="microsoft.ebpf-for-windows.bpf2c" type="win32" version="0.9.0.0"></assemblyIdentity>
  <asmv3:application>
    <asmv3:windowsSettings xmlns="http://schemas.microsoft.com/SMI/2020/WindowsSettings">
      <heapType>SegmentHeap</heapType>
    </asmv3:windowsSettings>
  </asmv3:application>
</assembly>

C:\ebpf-2023\ebpf-for-windows\x64\Debug>
```

```
C:\ebpf-2023\ebpf-for-windows>cd x64\FuzzerDebug

C:\ebpf-2023\ebpf-for-windows\x64\FuzzerDebug>type bpf2c.exe.manifest
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<!--
  Copyright (c) Microsoft Corporation
  SPDX-License-Identifier: MIT

  Overrides the default heap implementation with segment heap.
-->
<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0" xmlns:asmv3="urn:schemas-microsoft-com:asm.v3">
  <assemblyIdentity name="microsoft.ebpf-for-windows.bpf2c" type="win32" version="0.9.0.0"></assemblyIdentity>
  <asmv3:application>
    <asmv3:windowsSettings xmlns="http://schemas.microsoft.com/SMI/2020/WindowsSettings">
      <heapType>SegmentHeap</heapType>
    </asmv3:windowsSettings>
  </asmv3:application>
</assembly>

C:\ebpf-2023\ebpf-for-windows\x64\FuzzerDebug>
```
## Documentation

_Is there any documentation impact for this change? No

### CODE REVIEW

File: tools/bpf2c/bpf2c.exe.manifest [hunk 1]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<!--
+  Copyright (c) Microsoft Corporation
+  SPDX-License-Identifier: MIT
+
+  Overrides the default heap implementation with segment heap.
+-->
+<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0" xmlns:asmv3="urn:schemas-microsoft-com:asm.v3">
+  <assemblyIdentity name="bpf2c" type="win32" version="1.0.0.0"></assemblyIdentity>
```
Comment (Alan-Jowett) [inline tools/bpf2c/bpf2c.exe.manifest:9]: Can you please set the assemblyIdentity's name and version correctly? The version should match the semantic versioning of the project (i.e. 0.9.0 currently).

Comment (shpalani) [reply tools/bpf2c/bpf2c.exe.manifest]: Fixed. Thanks @Alan-Jowett

File: tools/bpf2c/bpf2c.exe.manifest [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<!--
+  Copyright (c) Microsoft Corporation
+  SPDX-License-Identifier: MIT
+
+  Overrides the default heap implementation with segment heap.
+-->
+<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0" xmlns:asmv3="urn:schemas-microsoft-com:asm.v3">
+  <assemblyIdentity name="microsoft.ebpf-for-windows.bpf2c" type="win32" version="0.9.0.0"></assemblyIdentity>
```
Comment (dthaler) [inline tools/bpf2c/bpf2c.exe.manifest:9]: If the 0.9.0 here is supposed to match the release version (meaning it should be updated each release), then you need to also update either `scripts\update-release-version.ps1` or `docs\ReleaseProcess.md`.  Check with @gtrevi to determine which.

Comment (shpalani) [reply tools/bpf2c/bpf2c.exe.manifest:9]: As per my understanding, yes, it is application binary version changes.

https://learn.microsoft.com/en-us/windows/win32/sbscs/assembly-versions

**Review by dthaler — CHANGES_REQUESTED** [review]
Waiting for either update-release-version.ps1 or ReleaseProcess.md to be updated in this PR.

File: tools/bpf2c/bpf2c.exe.manifest (hunk 2 above)
Comment (gtrevi) [reply tools/bpf2c/bpf2c.exe.manifest:9]: I agree with @dthaler, `scripts\update-release-version.ps1` should be updated for automating the version within this file, similarly to how it's done for the `Product.wxs` file (i.e. using regex, see lines 20-24). In this case, being the AssemblyVersion's standard made of four digits, I'd just keep the last one to `.0`.

The `docs\ReleaseProcess.md` shouldn't instead need any update, as the script was intended to remove most of the manual steps for releasing eBPF.

**Review by gtrevi — CHANGES_REQUESTED** [review]

File: tools/bpf2c/bpf2c.exe.manifest (hunk 2 above)
Comment (shpalani) [reply tools/bpf2c/bpf2c.exe.manifest:9]: @Alan-Jowett : https://github.com/microsoft/ebpf-for-windows/blob/main/scripts/Set-Version.ps1 gives
$version = "$major_version.$minor_version.$revision_version" (0.9.0)
but the requirement for the manifest version field needs a **mandatory** 4-digit version (0.9.0.0)

Comment (Alan-Jowett) [reply tools/bpf2c/bpf2c.exe.manifest:9]: I would suggest leaving the last part as 0 then. I.e. assume major=x, minor=y, revision=z, then the version in the manifest would be x.y.z.0.

Comment (shpalani) [reply tools/bpf2c/bpf2c.exe.manifest]: Thanks @Alan-Jowett. As per the suggestion, made {version} to be populated in the bpf2c.exe.manifest by Set-Version.ps1 execution.

**Comment by shpalani** [comment]
> Waiting for either update-release-version.ps1 or ReleaseProcess.md to be updated in this PR.

Changed the approach to have input file bpf2c.exe.manifest.in with the {version} param to be replaced and generate the output file bpf2c.exe.manifest. This generated manifest file will be embedded in the app/bfp2c binary.

**Review by Alan-Jowett — APPROVED** [review]

**Review by dthaler — APPROVED** [review]

**Pull request merged**